lib/smooth/core/io/UART.cpp

```diff
@@ -0,0 +1,574 @@
+/*
+Smooth - A C++ framework for embedded programming on top of Espressif's ESP-IDF
+Copyright 2019 Per Malmberg (https://gitbub.com/PerMalmberg)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <sstream>
+
+#include "smooth/core/io/UART.h"
+#include "smooth/core/logging/log.h"
+
+using namespace smooth::core::logging;
+
+namespace smooth::core::io
+{
+  constexpr const char* log_tag = "UART";
+
+  UART::UART(uart_port_t port, int baud_rate, uart_word_length_t data_bits, uart_parity_t parity, uart_stop_bits_t stop_bits, uart_hw_flowcontrol_t flow_control, gpio_num_t tx_pin, gpio_num_t rx_pin, gpio_num_t rts_pin, gpio_num_t cts_pin, uart_mode_t mode)
+      : port_(port)
+  {
+    this->config_ = uart_config_t{
```

Comment:
No need for `this->` and also from The contributing.md document: 
Do not use Hungarian notation for variables, i.e. warts m_